rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java

```diff
@@ -239,13 +247,187 @@ public void traceBlock(
+        }
+
+        public synchronized void addPaidFees(Coin paidFees) {
+            totalPaidFees = totalPaidFees.add(paidFees);
+        }
+
+        public synchronized void addGasUsed(long gasUsed) {
+            this.totalGasUsed += gasUsed;
+        }
+
+        public synchronized int addReceipt(TransactionReceipt receipt) {
+            if (receipts.add(receipt)) {
+                return receipts.size() - 1;
+            }
+            return -1;
+        }
+
+        public synchronized void addExecutedTransaction(Transaction tx) {
+            executedTransactions.add(tx);
+        }
+
+        public synchronized void addDeletedAccounts(Collection<DataWord> deletedAccountsToAdd) {
+            deletedAccounts.addAll(deletedAccountsToAdd);
+        }
+
+        public Coin getTotalPaidFees() {
+            return totalPaidFees;
+        }
+
+        public long getTotalGasUsed() {
+            return totalGasUsed;
+        }
+
+        public List<TransactionReceipt> getReceipts() {
+            return new ArrayList<>(receipts);
+        }
+
+        public List<Transaction> getExecutedTransactions() {
+            return new ArrayList<>(executedTransactions);
+        }
+
+        public Set<DataWord> getDeletedAccounts() {
+            return new HashSet<>(deletedAccounts);
+        }
+
+        public Repository getRepository() {
+            return track;
+        }
+
+        public ProgramTraceProcessor getProgramTraceProcessor() {
+            return programTraceProcessor;
+        }
+
+    }
+
+    public static class TransactionConflictException extends Exception {
+        public TransactionConflictException(String message) {
+            super(message);
+        }
+    }
+
+    /**
+     * Pre-runs each transaction to create the partitioning in different thread, assuming that there shall be
+     * no conflicts between transactions assigned to different partitions (= will be run in different threads)
+     *
+     * @param block
+     * @param track
+     * @param programTraceProcessor
+     * @param acceptInvalidTransactions
+     * @param discardInvalidTxs
+     * @return true if partitioning has succeed. Returning false means the block can't be validated
+     */
+    private boolean computeTxPartitioning(Block block, RepositorySnapshot track, ProgramTraceProcessor programTraceProcessor,
+                                          boolean acceptInvalidTransactions, boolean discardInvalidTxs) {
+        //
+        // Work with a cached repository that will be rolled back in the end of transaction execution
+        TransactionsPartitioner partitioner = new TransactionsPartitioner();
+        Repository dummyRepo = track.startTracking();
+        BlockSharedData dummySharedData = new BlockSharedData(dummyRepo, programTraceProcessor);
+        TransactionConflictDetectorWithCache transactionConflictDetector = new TransactionConflictDetectorWithCache(partitioner);
+        if (dummyRepo.isCached()) {
```

Comment:
I don't have enough clarity why the TransactionConflictDetectorWithCache is used here and not the TransactionConflictDetector.

Reply:
During block mining, in case a transaction is invalid, we actually have started to execute it. All the state changes will be rollbacked thanks to the cache system at the repository level. Nevertheless, the conflict detector has recorded the accesses to the state keys from this transaction execution. We also need to purge the access when discarding the tx from the block. Then, we have 2 options :
1.	We can record accesses to state keys from the partition executing the transaction and, if it is discarded, we purge the entries from that partition in internal maps. But this reverse mechanism is not really efficient because the accesses are mapped ‘per key’, not ‘per partition’ (for performance reason in the nominal cases).
2	Or, we can cache the accesses to state keys from the partition executing the transaction and, if it is NOT discarded, we merge the entries from that partition in internal maps.

I’ve started with implementing the 1st option, successfully, but I was not happy with the resulting purging code, not really efficient and a too tricky for maintenance.

Then, I’ve decided to refactor towards the second option using a derivated class of TransactionConflictDetector that caches the entries until the commitPartition() method is called (merge implementation). Calling commitPartition() is only called if the transaction is valid, otherwise, we just clear the cache to ignore all accesses from the associated partition. 

Ultimately, TransactionConflictDetector is still used for block validation (executing txs in parallel regarding the partitioning) to detect conflicts and immediately stop block execution if such a case happens, whereas TransactionConflictDetectorWithCache is used for block mining.

I’ve added a comment to describe the TransactionConflictDetectorWithCache class and explain what this class is used for.